Server/controllers/authController.js

```diff
@@ -261,6 +262,73 @@ const resetPasswordController = async (req, res, next) => {
     next(error);
   }
 };
+
+/**
+ * Deletes a user and all associated monitors, checks, and alerts.
+ *
+ * @param {Object} req - The request object.
+ * @param {Object} res - The response object.
+ * @param {Function} next - The next middleware function.
+ * @returns {Object} The response object with success status and message.
+ * @throws {Error} If user validation fails or user is not found in the database.
+ */
+const deleteUserController = async (req, res, next) => {
+  try {
+    // Validate user
+    await deleteUserParamValidation.validateAsync(req.body);
+
+    // Check if the user exists
+    const user = await req.db.getUserByEmail(req);
```

Comment:
There's a bug here, what happens if a user uses their token maliciously and adds a different user's email to their request?

The malicious use will be able to delete any other user's monitors. 

I think to resolve this we need to extract the user's ID and Email from their token rather than allowing it to be passed in the body or request params

Reply:
So, what do you have in mind for it to added?
Could you please explain your idea a bit more with a code snippet?

Reply:
Sure, when the user hits the `delete` endpoint they must have a valid JWT token.  This is included in the headers:

```
{
  host: 'localhost:5000',
  'user-agent': 'insomnia/2023.5.8',
  authorization: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI2NjdkZWQxOGFlNjExNDFlNjg5NzQ0NjQiLCJmaXJzdG5hbWUiOiJBbGV4IiwibGFzdG5hbWUiOiJIb2xsaWRheSIsImVtYWlsIjoidGVzdEB0ZXN0LmNvbSIsImlzQWN0aXZlIjp0cnVlLCJpc1ZlcmlmaWVkIjpmYWxzZSwiY3JlYXRlZEF0IjoiMjAyNC0wNi0yN1QyMjo1MjowOC44MDBaIiwidXBkYXRlZEF0IjoiMjAyNC0wNi0yN1QyMjo1MjowOC44MDBaIiwiX192IjowLCJpYXQiOjE3MTk1Mjg5OTUsImV4cCI6MTcyODA4MjU5NX0.FBd_TmyCfMHKpLm_1cyzgcpeDLIZaT14XkkNaOka1LQ',
  accept: '*/*'
}
```

The JWT token contains both the user ID and the user's email.  Decoded JWT:

```
{
  "_id": "667ded18ae61141e68974464",
  "firstname": "Alex",
  "lastname": "Holliday",
  "email": "test@test.com",
  "isActive": true,
  "isVerified": false,
  "createdAt": "2024-06-27T22:52:08.800Z",
  "updatedAt": "2024-06-27T22:52:08.800Z",
  "__v": 0,
  "iat": 1719528995,
  "exp": 1728082595
}
```

So my suggestion is that we remove all parameters and the request body from this request and obtain all information we need from the JWT.

That way a user cannot maliciously affect another user's data unless they have stolen their token.  That's beyond the scope of our concern though, there's not much we can do about that other than set very short expirty times for the token.

In summary, my suggestion is:

1.  Remove the request parameter and request body
2.  Get the JWT token from `req.headers`
3.  Decode the JWT and extract the user ID and email.  We already have the `jsonwebtoken` library included for decoding tokens, see `verifyJWT` middleware for usage example
4.  Continue as normal for the rest of the process.

I think this gives us the greatest amount of security for a very sensitive endpoint

Reply:
Oh, got it 👍🏻
I've seen this before!

Reply:
I tried to implement your idea.
Check to see if that is what you mean